pkg/migrator/worker.go

```diff
@@ -0,0 +1,701 @@
+						"getting next batch of records",
+						zap.Int64(lastMigratedIDField, lastMigratedID),
+					)
+
+					records, err := metrics.MeasureReaderLatency(
+						w.tableName,
+						func() ([]ISourceRecord, error) {
+							return reader.Fetch(ctx, lastMigratedID, w.batchSize)
+						},
+					)
+					if err != nil {
+						switch err {
+						case sql.ErrNoRows:
+							logger.Info(noMoreRecordsToMigrateMessage)
+
+							metrics.EmitMigratorReaderNumRowsFound(w.tableName, 0)
+
+							select {
+							case <-ctx.Done():
+								return
+							case <-time.After(sleepTimeOnNoRows):
+							}
+
+						default:
+							metrics.EmitMigratorReaderError(w.tableName, err.Error())
+
+							logger.Error(
+								"getting next batch of records failed, retrying",
+								zap.Error(err),
+							)
+
+							select {
+							case <-ctx.Done():
+								return
+							case <-time.After(sleepTimeOnError):
+							}
+						}
+
+						continue
+					}
+
+					metrics.EmitMigratorReaderNumRowsFound(w.tableName, int64(len(records)))
+
+					if len(records) == 0 {
+						logger.Info(noMoreRecordsToMigrateMessage)
+
+						select {
+						case <-ctx.Done():
+							return
+						case <-time.After(sleepTimeOnNoRows):
+						}
+
+						continue
+					}
+
+					for _, record := range records {
+						id := record.GetID()
+
+						_, seen := w.isInflight(id)
+
+						if seen {
+							continue
+						}
+
+						select {
+						case <-ctx.Done():
+							logger.Info(contextCancelledMessage)
+							return
+						case <-w.sem:
+						}
+
+						select {
+						case <-ctx.Done():
+							logger.Info(contextCancelledMessage)
+							return
+
+						case w.recvChan <- record:
+							w.addInflightStartTime(id, startE2ELatency)
+
+							logger.Debug(
+								"sent record to transformer",
+								zap.Int64(idField, id),
+							)
+						}
+					}
+				}
+			}
+		})
+
+	return nil
+}
+
+func (w *Worker) startTransformer(ctx context.Context, transformer IDataTransformer) error {
+	logger := w.logger.Named(utils.MigratorTransformerLoggerName).
+		With(zap.String(tableField, w.tableName))
+
+	logger.Info("started")
+
+	tracing.GoPanicWrap(
+		ctx,
+		&w.wg,
+		fmt.Sprintf("transformer-%s", w.tableName),
+		func(ctx context.Context) {
+			defer close(w.wrtrChan)
+
+			for {
+				select {
+				case <-ctx.Done():
+					logger.Info(contextCancelledMessage)
+					return
+
+				case record, open := <-w.recvChan:
+					if !open {
+						logger.Info(channelClosedMessage)
+						return
+					}
+
+					envelope, err := transformer.Transform(record)
+					if err != nil {
+						logger.Error(
+							"failed to transform",
+							zap.Error(err),
+							zap.Int64(idField, record.GetID()),
+						)
+
+						err := insertMigrationDeadLetterBox(
+							ctx,
+							w.writer,
+							w.tableName,
+							record.GetID(),
+							nil,
+							FailureTransformerError,
+						)
+						if err != nil {
+							logger.Error("failed to insert dead letter box", zap.Error(err))
+						}
```

Comment:
should this retry or continue?

Reply:
Transformer errors are basically "something really wrong happens with the source payload", so I'd default to not retry these. Also, I don't expect to have this kind of dead letter.